### DISCUSSION TITLE
micropython_ir with A1156 Apple Remote

### DISCUSSION THREAD

**Discussion by Pax-nl** [micropython/micropython#13327] [Libraries & Drivers]
I'm trying to use an old Apple Remote with micropython_ir.

It seems for lircd Jeremy found the following:

```
# hacked by Jeremy Kidwell
#
# brand:                       Apple
# model no. of remote control: A1156
# devices being controlled by this remote: new (late 2006) MacBook
#
# This config files are for non-Apple receivers only.
# Use the lircd.conf.macmini file when you are using the Apple receiver.
#

begin remote
  name  Apple_A1156
  bits            8
  flags   SPACE_ENC
  eps            30
  aeps          100

  header       9065  4484
  one           574  1668
  zero          574   547
  ptrail        567
  repeat       9031  2242
  pre_data_bits   16
  pre_data       0x77E1
  post_data_bits  8
  post_data      0xC5
  gap          37600
  toggle_bit      0
  ignore_mask 0x80ff

      begin codes
          KEY_PAUSE                0x20
	  KEY_VOLUMEDOWN		   0xB0
          KEY_VOLUMEUP 	 	 	   0xD0
          KEY_NEXT                 0xE0
          KEY_PREVIOUS             0x10
      end codes

end remote
```

Is it possible to get this remote working with micropython_ir?

**Comment by peterhinch** [comment]
I think there is sufficient information to do this but (as author of micropython-ir) I would be unable to undertake this as I have no device controlled by such a remote. If you feel able to tackle this I'll do my best to advise. The basic timings match the popular NEC protocol so I would start by adapting that driver. [NEC](https://www.sbprojects.net/knowledge/ir/nec.php) sends address followed by its 1s complement, then data followed by its complement. Apple seems to send a preamble and a post-amble. It's not clear to me from the above whether it sends an address and whether complements are used.

If I had an Apple remote I would examine the demodulated waveform with a scope or logic analyser to settle these questions.

Good luck.

**Reply by Pax-nl** [reply]
Thank you, a scope is something that is still on my wish list.

f.y.i I'm not interested in controlling a device that is compatible with the remote, I want to use the remote to eventually send mqtt messages. The remote was laying around and I thought this is a very nice small device to control my Sonos speakers. But until now there is no out of the box solution for Micro or Circuit Python. 

I'm currently looking into this Arduino library: https://github.com/crankyoldgit/IRremoteESP8266 which has some support for the remote: https://github.com/crankyoldgit/IRremoteESP8266/pull/265

But as I'm moving away from Arduino, I really wanted to get it working with micropython. 2 projects are still on Arduino. There was no python library and at the time and I could not write my own, but this looks promising: https://github.com/erikdelange/ITHO-CVU-controller-in-MicroPython no need to re-create the wheel sort of speak :D

I digress, if the Arduino library does not work I will return to your library and try to get it working with the hints you gave, again thank you for the quick response.
This can be helpful too: https://en.wikipedia.org/wiki/Apple_Remote

**Reply by Pax-nl** [reply]
@peterhinch 

I forked your repository and created a feature branch: https://github.com/Pax-IT/micropython_ir/tree/feature/apple_remote

I need some help with:

- double callbacks, probably repeat code issue
- command does not have parity? found this in ir_NEC.hpp (Arduino-IRremote)
```
// The parity is not implemented, so we get: <start bit><0x87EE:16><device ID:8><command:8><stop bit>
```
- Virtual Timer(-1) issue: https://github.com/peterhinch/micropython_ir/issues/35

For now I got what I need and can continue, but if you want me to create a pull request I need some help with the issues I mentioned.

Added an example for the Apple Remote in examples/apple/main.py

**Reply by peterhinch** [reply]
I'm sorry but I really can't see how I can help with the Apple remote. I don't have a sample to test and have no knowledge of the device. I'd have to websearch for information, then convey it to you. Quicker for you to cut out the middleman.

Re the crash, I'll wait a few days to see if there's a response from the maintainers. If there is none I'll push a patch that checks for a C3 and switches to timer 0 (or some similar hack).

I'm not keen on a PR. In the years since I wrote this, you are the first to want to use this remote. Also I don't like being in a position of having to support hardware I don't possess. If you get a working solution, by all means post a link to your fork and I'll mention it in the docs.

**Comment by Pax-nl** [comment]
Thanks.. I will close the question. 

(the link to the fork is in my previous message)

**Reply by peterhinch** [reply]
Please let me know when your fork is fully working and I'll link to it in the docs.

**Reply by Pax-nl** [reply]
@peterhinch 

Is it ok if I use your `ir_rx/__init__.py` in my own repo? This way I can publish a working version of the receiver and you do not have to maintain anything?

**Reply by peterhinch** [reply]
Of course. I suggest you keep the directory structure and contents.
